src/github_runner_image_builder/builder.py

```diff
@@ -0,0 +1,793 @@
+    """
+    # The commands will fail if artefacts do not exist and hence there is no need to check the
+    # output of subprocess runs.
+    try:
+        output = subprocess.run(
+            ["/usr/bin/umount", str(IMAGE_MOUNT_DIR / "dev")],
+            timeout=30,
+            check=False,
+        )  # nosec: B603
+        logger.info("umount dev out: %s", output)
+        output = subprocess.run(
+            ["/usr/bin/umount", str(IMAGE_MOUNT_DIR / "proc")],
+            timeout=30,
+            check=False,
+            capture_output=True,
+        )  # nosec: B603
+        logger.info("umount proc out: %s", output)
+        output = subprocess.run(
+            ["/usr/bin/umount", str(IMAGE_MOUNT_DIR / "sys")],
+            timeout=30,
+            check=False,
+            capture_output=True,
+        )  # nosec: B603
+        logger.info("umount sys out: %s", output)
+        output = subprocess.run(
+            ["/usr/bin/umount", str(IMAGE_MOUNT_DIR)], timeout=30, check=False, capture_output=True
+        )  # nosec: B603
+        logger.info("umount ubuntu-image out: %s", output)
+        output = subprocess.run(
+            ["/usr/bin/umount", str(NETWORK_BLOCK_DEVICE_PATH)],
+            timeout=30,
+            check=False,
+            capture_output=True,
+        )  # nosec: B603
+        logger.info("umount nbd out: %s", output)
+        output = subprocess.run(  # nosec: B603
+            ["/usr/bin/umount", str(NETWORK_BLOCK_DEVICE_PARTITION_PATH)],
+            timeout=30,
+            check=False,
+            capture_output=True,
+        )
+        logger.info("umount nbdp1 out: %s", output)
+    except subprocess.SubprocessError as exc:
+        raise UnmountBuildPathError from exc
+
+
+def _download_and_validate_image(arch: Arch, base_image: BaseImage) -> Path:
+    """Download and verify the base image from cloud-images.ubuntu.com.
+
+    Args:
+        arch: The base image architecture to download.
+        base_image: The ubuntu base image OS to download.
+
+    Returns:
+        The downloaded image path.
+
+    Raises:
+        BaseImageDownloadError: If there was an error with downloading/verifying the image.
+    """
+    try:
+        bin_arch = _get_supported_runner_arch(arch)
+    except UnsupportedArchitectureError as exc:
+        raise BaseImageDownloadError from exc
+
+    image_path_str = f"{base_image.value}-server-cloudimg-{bin_arch}.img"
+    image_path = _download_base_image(
+        base_image=base_image, bin_arch=bin_arch, output_filename=image_path_str
+    )
+    shasums = _fetch_shasums(base_image=base_image)
+    if image_path_str not in shasums:
+        raise BaseImageDownloadError("Corresponding checksum not found.")
+    if not _validate_checksum(image_path, shasums[image_path_str]):
+        raise BaseImageDownloadError("Invalid checksum.")
+    return image_path
+
+
+def _get_supported_runner_arch(arch: Arch) -> SupportedBaseImageArch:
```

Comment:
Can you just try to instantiate whatever string you get as an enum containing valid values for this?

Reply:
The values returned by `platform.machine()` and the values used by `cloud-images` and values used by Openstack are all different (e.g. `aarch64` vs `arm64` vs `aarch`) and hence I had to create own supported_runner_arch conversion method for each services.